randomization/distribution_sample_generator.cc

```diff
@@ -0,0 +1,142 @@
+  float cdf = 0;
+  for (const auto& point : config.pmf_points()) {
+    cdf += point.pmf();
+  }
+  if (cdf != 1) {
+    return absl::InvalidArgumentError(
+        absl::StrCat("Cumulative value of all PMFs should be 1. It is '", cdf,
+                     "' instead."));
+  }
+  return absl::OkStatus();
+};
+
+absl::Status ValidateCdfConfig(DistributionConfig& config) {
+  if (config.cdf_points_size() == 0) {
+    return absl::InvalidArgumentError(
+        absl::StrCat("CDF is not provided for '", config.name(), "'."));
+  }
+
+  auto prev_cdf = config.cdf_points(0).cdf();
+  if (prev_cdf < 0) {
+    return absl::InvalidArgumentError(
+        absl::StrCat("The cdf value:'", prev_cdf,
+                     "' must not be negative in CDF:'", config.name(), "'."));
+  }
+
+  auto prev_value = config.cdf_points(0).value();
+  for (int i = 1; i < config.cdf_points_size(); i++) {
+    auto curr_value = config.cdf_points(i).value();
+    auto curr_cdf = config.cdf_points(i).cdf();
+    if (curr_value <= prev_value) {
+      return absl::InvalidArgumentError(absl::StrCat(
+          "The value:'", curr_value, "' must be greater than previous_value:'",
+          prev_value, "' at index '", i, "' in CDF:'", config.name(), "'."));
+    }
+    if (curr_cdf < prev_cdf) {
+      return absl::InvalidArgumentError(
+          absl::StrCat("The cdf value:'", curr_cdf,
+                       "' must be greater than previous cdf value:'", prev_cdf,
+                       "' at index '", i, "' in CDF:'", config.name(), "'."));
+    }
+    prev_value = curr_value;
+    prev_cdf = curr_cdf;
+  }
+
+  auto last_configured_cdf =
+      config.cdf_points(config.cdf_points_size() - 1).cdf();
+  if (last_configured_cdf != 1) {
+    return absl::InvalidArgumentError(absl::StrCat(
+        "The maximum value of cdf:'", last_configured_cdf, "' in CDF:'",
+        config.name(), "' must be exactly equal to 1."));
+  }
+  return absl::OkStatus();
+};
+
+absl::Status ValidateDistributionConfig(DistributionConfig& config) {
+  auto cdf_present = config.cdf_points_size() != 0;
+  auto pmf_present = config.pmf_points_size() != 0;
+
+  if (cdf_present == pmf_present) {
+    return absl::InvalidArgumentError(
+        absl::StrCat("Exactly one of CDF and PMF must be provided for '",
+                     config.name(), "'."));
+  }
+
+  if (cdf_present) return ValidateCdfConfig(config);
+  if (pmf_present) return ValidatePmfConfig(config);
+
+  return absl::InvalidArgumentError(
+      absl::StrCat("Review CDF and PMF for '", config.name(), "'."));
+};
+
+std::unique_ptr<DistributionSampleGenerator> AllocateSampleGeneratorLibrary(
+    const DistributionConfig& config) {
+  std::unique_ptr<DistributionSampleGenerator> sample_gen_lib;
+  sample_gen_lib = std::make_unique<DistributionSampleGenerator>();
+  sample_gen_lib->Initialize(config);
+  return sample_gen_lib;
+};
+
+void DistributionSampleGenerator::InitializeWithCdf(
```

Comment:
what is this function trying to do? Is it doing the right thing?

Reply:
I added a call to validate config inside Initialize(). Also changed the type of InitialzePmf/Cdf to absl::Status so it will FAIL for wrong configs.